subsystems/mvdm/ntvdm/dos/dos32krnl/dosfiles.c

```diff
@@ -1125,11 +1125,79 @@ BOOLEAN DosUnlockFile(WORD DosHandle, DWORD Offset, DWORD Size)
     return TRUE;
 }
 
+BOOLEAN DosDeviceIoControlDrive(WORD DriveNumber, BYTE ControlCode, DWORD Buffer, PWORD Result)
+{
+    CHAR RootPath[] = "?:\\";
+
+    if (DriveNumber == 0x00)
+        RootPath[0] = 'A' + Sda->CurrentDrive;
+    else
+        RootPath[0] = 'A' + DriveNumber - 1;
+
+    switch (ControlCode)
+    {
+        case 0x04:
+            DPRINT1("UNIMPLEMENTED INT 21h, 4404h, Read from block device %s\n", RootPath);
+            Sda->LastErrorCode = ERROR_INVALID_FUNCTION;
+            break;
+        case 0x05:
+            DPRINT1("UNIMPLEMENTED INT 21h, 4405h, Write block device control string %s\n", RootPath);
+            Sda->LastErrorCode = ERROR_INVALID_FUNCTION;
+            break;
+        case 0x08:
+        {
+            DWORD DriveType = GetDriveTypeA(RootPath);
+
+            switch (DriveType)
+            {
+            case DRIVE_UNKNOWN:
+            case DRIVE_NO_ROOT_DIR:
+            default:
+                DPRINT1("INT 21h, 4408h, %s -> DriveType = 0x%x\n", RootPath, DriveType);
+                *Result = 0x000f;
+                return TRUE;
+            case DRIVE_REMOVABLE:
+            case DRIVE_CDROM:
+                *Result = 0x0000;
+                return TRUE;
+            case DRIVE_FIXED:
+                *Result = 0x0001;
+                return TRUE;
+            case DRIVE_REMOTE:
+            case DRIVE_RAMDISK: // ??
+                break;
```

Comment:
Maybe these two should also fall back to the "default:" case?

Reply:
No clue, and I have no environment to test it.